File: new-post.js
'use strict';

const path = require('path');
const fs = require("fs");
const execSync = require('child_process').execSync;

// レイアウト
const LAYOUT = 'page';
// 新しいポストを作成するパス
const OUTPUT_PATH = ['_source', '_posts'];
// 拡張子
const EXTENSION = '.md';
// 新しいPostを開くエディターのパス
const EDITOR = 'code';
// ファイル名に使用できない文字 '!"#$%&()*,/:;?@[]^`{|}\+<=>
const INVALID_FILENAME_CHAR = /[ '!"#\$%&\(\)\*,\/:;\?@\[\]\^`\{\|\}\\\+<=>]/g;

//------------------------------------------------------------------------------
// 日付文字列を作成
//------------------------------------------------------------------------------
function createDateString(date) {
    const YYYY = date.getFullYear();
    const MM = (date.getMonth() + 1).toString().padStart(2, '0');
    const DD = date.getDate().toString().padStart(2, '0');

    return `${YYYY}-${MM}-${DD}`;
}

//------------------------------------------------------------------------------
// 時刻文字列を作成
//------------------------------------------------------------------------------
function createTimeString(date) {
    const hh = date.getHours().toString().padStart(2, '0');
    const mm = date.getMinutes().toString().padStart(2, '0');
    const ss = date.getSeconds().toString().padStart(2, '0');

    return `${hh}:${mm}:${ss}`;
}

//------------------------------------------------------------------------------
// YAML front matterを作成
//------------------------------------------------------------------------------
function createYAMLFrontMatter(date, title) {
    const dateTimeString = createDateString(date) + ' ' + createTimeString(date);

    return (
`---
layout: ${LAYOUT}
title: ${title}
date: ${dateTimeString} +0900
category: blog
tags: []
description:
---

`
    );
}

//-----------------------------------------------------------------------------
// ファイル名を作成
//-----------------------------------------------------------------------------
function createFileName(date, title) {
    const dateString = createDateString(date);
    const fileName = title.replace(INVALID_FILENAME_CHAR, '-').toLowerCase();
    return `${dateString}-${fileName}${EXTENSION}`;
}

//-----------------------------------------------------------------------------
// 新しいPostを作成
//-----------------------------------------------------------------------------
function createNewPost(title) {
    const today = new Date();
    const frontMatter = createYAMLFrontMatter(today, title);
    const fileName = createFileName(today, title);
    const filePath = path.join(__dirname, ...OUTPUT_PATH, fileName);

    fs.writeFileSync(filePath, frontMatter);

    return filePath;
}

//-----------------------------------------------------------------------------
// メイン
//-----------------------------------------------------------------------------
function main() {
    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    // カレントディレクトリをスクリプトのあるディレクトリへ変更
    process.chdir(__dirname);

    readline.question('タイトルを入力して下さい: ', title => {
        readline.close();

        const filePath = createNewPost(title);
        execSync(`${EDITOR} ${filePath}`);
    });
}

main();

File: new-post.py
from pathlib import Path
import re
import subprocess
import datetime
import os


# レイアウト
LAYOUT = "page"
# 新しいポストを作成するパス
OUTPUT_PATH = ["_source", "_posts"]
#  拡張子
EXTENSION = ".md"
#  新しいPostを開くエディターのパス
EDITOR = "code"
#  ファイル名に使用できない文字 '!"#$%&()*,/:;?@[]^`{|}\+<=>
INVALID_FILENAME_CHAR = re.compile(r"[ '!\"#$%&\(\)\*,/:;?@\[\]^`{|}\\+<=>]+")


# ------------------------------------------------------------------------------
# YAML front matterを作成
# ------------------------------------------------------------------------------
def create_yaml_front_matter(date: datetime.datetime, title: str) -> str:
    date_time_string = date.strftime("%Y-%m-%d %H:%M:%S")
    return f"""---
layout: {LAYOUT}
title: {title}
date: {date_time_string} +0900
category: blog
tags: []
description:
---

"""


# -----------------------------------------------------------------------------
# ファイル名を作成
# -----------------------------------------------------------------------------
def create_file_name(date: datetime.datetime, title: str) -> str:
    date_string = date.strftime("%Y-%m-%d")
    sanitized_title = INVALID_FILENAME_CHAR.sub('-', title).lower()
    return f"{date_string}-{sanitized_title}{EXTENSION}"


# -----------------------------------------------------------------------------
# 新しいPostを作成
# -----------------------------------------------------------------------------
def create_new_post(title: str) -> Path:
    today = datetime.datetime.now()
    front_matter = create_yaml_front_matter(today, title)
    output_path = os.path.join(Path(__file__).parent, *OUTPUT_PATH)
    os.makedirs(output_path, exist_ok=True)

    file_name = create_file_name(today, title)
    file_path = os.path.join(output_path, file_name)

    with open(file_path, "w", encoding="utf-8") as f:
        f.write(front_matter)

    return file_path


# -----------------------------------------------------------------------------
# メイン
# -----------------------------------------------------------------------------
def main():
    title = input("タイトルを入力して下さい: ")
    file_path = create_new_post(title)
    subprocess.run([EDITOR, file_path])


if __name__ == "__main__":
    main()
